pynbody/halo/adaptahop.py

```diff
@@ -165,11 +178,75 @@ def _read_halo_data(self, halo_id, offset):
         props['members'] = iord_array
 
         # Create halo object and fill properties
-        halo = Halo(halo_id, self, self.base.dm, index_array=None, iord_array=iord_array)
+        if hasattr(self, '_group_to_indices'):
+            index_array = self._group_to_indices[halo_id]
+            iord_array = None
+        else:
+            index_array = None
+            iord_array = iord_array
+        halo = Halo(halo_id, self, self._base_dm, index_array=index_array, iord_array=iord_array)
         halo.properties.update(props)
 
         return halo
 
+    def get_group_array(self, family='dm', group_to_indices=False):
+        """Return an array with an integer for each particle in the simulation
+        indicating which halo that particle is associated with. If there are multiple
+        levels (i.e. subhalos), the number returned corresponds to the lowest level, i.e.
+        the smallest subhalo.
+
+        Arguments
+        ---------
+        family : optional, default : dm
+            The family of the particles that make the group
+        group_to_indices : optional, bool
+            If True, store the mapping from groups to particle on-disk location.
+
+        Returns
+        -------
+        igrp : int array
+            An array that contains the index of the group that contains each particle.
+        """
+        logger.debug('Get_group_array')
+        if family is None:
+            family == self.base.families()[0]
+        elif isinstance(family, str):
+            families = self.base.families()
+            matched_families = [f for f in families if f.name == family]
+            if len(matched_families) != 1:
+                raise Exception('Could not find family %s' % family)
+            family = matched_families[0]
+        try:
+            data = self.base[family]
+        except:
+            logger.error((type(self.base)))
+            logger.error((type(family)))
+            raise
+
+        iord = data['iord']
+        iord_argsort = data['iord_argsort']
+
+        igrp = np.zeros(len(data), dtype=int) - 1
+
+        if group_to_indices:
+            grp2indices = {}
+        with FortranFile(self._fname) as fpu:
+            for halo_id, halo in self._halos.items():
+                fpu.seek(halo.properties['file_offset'])
+                fpu.skip(1)  # number of particles
+                particle_ids = fpu.read_vector('i')
```

Comment:
After #584 is merged, this should be fixed to take into account cases where `particle_ids` is not sorted and/or is not a 32bit integer.